[feat][doc] add conceptual docs (bundle assignment/splitting/unloading...) for broker load balancing (#644)
apache/pulsar-site: docs/concepts-broker-load-balancing-concepts.md

sidebar_label: "Concepts"
---

````mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
````

> Pulsar provides robust support for load balancing to ensure efficient utilization of resources across Pulsar clusters. Load balancing in Pulsar involves distributing messages and partitions evenly among brokers and consumers to prevent hotspots and optimize performance.

Reviewer: ```suggestion
Pulsar provides robust load balancing to ensure efficient utilization of resources across Pulsar clusters. Load balancing in Pulsar involves distributing messages and partitions evenly among brokers and consumers to prevent hotspots and optimize performance.
```

Author: I'd rather to keep it since ` xx provides robust support for ...` is a commonly used phrase

Before getting started with load balancing, it's important to review the key components to ensure that resources are utilized efficiently and varying workloads can be handled by the system effectively.

## Brokers

In a Pulsar cluster, [brokers](./reference-terminology.md#broker) are responsible for serving messages for different topics and partitions. Broker load balancing ensures that each broker handles a proportional share of the load. 

## Producers

[Producers](./reference-terminology.md#producer) in Pulsar are responsible for publishing messages to topics. Pulsar clients (producers) connect to brokers to publish messages. Producer load balancing (i.e., connection pooling mechanism in Pulsar) ensures that producers are distributed across brokers to avoid overwhelming a single broker with too many connections.

## Consumers

[Consumers](./reference-terminology.md#consumer) in Pulsar are responsible for consuming messages from topics. Depending on how consumer load balancing is configured (i.e., using exclusive or shared consumers or auto-rebalancing), you can ensure even load distribution.

## Topics

> [Topics](./reference-terminology.md#topic) are the basic units for clients to publish and consume messages. Related topics are logically grouped into a namespace. To efficiently manage metadata and keep track of all of them moving through the system, Pulsar uses a strategy of grouping topics by partitioning on a namespace to create topic bundles.

Reviewer: ```suggestion
[Topics](./reference-terminology.md#topic) are the basic units for clients to publish and consume messages. Related topics are logically grouped into a namespace. To efficiently manage metadata and keep track of all of them moving through the system, Pulsar groups topics by partitioning on a namespace to create topic bundles.
```

Author: I'd like to keep the original text since it's a correct sentence, from both technical and English sides.

![Relationships of topics and namespaces](/assets/broker-load-balancing-1.png)

## Bundles

[Bundles](./reference-terminology.md#namespace-bundle) represent a range of partitions for a particular namespace in Pulsar, comprising a portion of the overall hash range of the namespace. 

> Bundle is introduced in Pulsar to represent a middle-layer group. Each bundle is an **assignment unit**, which means topics are assigned to brokers at the **bundle** level rather than the topic level.

Reviewer: ```suggestion
Bundles in Pulsar represent middle-layer groups. Each bundle is an **assignment unit**, which means topics are assigned to brokers at the **bundle** level rather than the topic level.
```

## Broker load balancing

> 

Reviewer: Broker load balancer? As you say "it's important to review the key components to ensure..." at the beginning. It seems the broker load balancer is more like a component than broker load balancing.

Author: I'd prefer to keep it as it is since broker load balancing is the feature while broker load balancer is a component

When properly balanced, the brokers can handle increased traffic and ensure that the system can scale seamlessly to accommodate growing workloads. Load balancing helps prevent bottlenecks and ensures that the resources of the cluster are utilized optimally, leading to better throughput and reduced message processing latency.

![Concept of broker load balancer](/assets/broker-load-balancing-2.png)

## Topic bundling 

Topic bundling refers to the process of grouping topics into bundles. Pulsar organizes topics into bundles within a namespace. Each bundle is a range of partitions, and Pulsar can automatically distribute these bundles across brokers to achieve load balancing. This allows the cluster to scale more efficiently as brokers can independently manage their assigned bundles.

For example, 

- Topic load statistics (e.g., message rates) are aggregated at the **bundle** layer, which reduces the cardinality of load samples to monitor.

> - For dynamic topic-broker assignments, Pulsar persists these mappings at the **bundle **level, which decreases the space for storing dynamic topic-broker ownerships.

Reviewer: at **bundle **level or at the **bundle** layer? Should keep consistent?

Author: ok

You can split bundles in the following ways:

- Automatic: enable Pulsar's automatic bundle splitting process when a namespace has a significant increase in workload or the number of partitions exceeds the optimal capacity for a single bundle.

- Manual: trigger bundle splitting manually, to divide an existing bundle into multiple smaller bundles. 

Bundle splitting methods|Definition|When to use
|---|---|---
Automatic|Bundles are split automatically based on different [bundle splitting algorithms](#bundle-splitting-algorithms). | Automatic bundle splitting is most commonly used.<br/><br/>You can use this method in various scenarios, such as when a bundle remains hot for a long time. 
Manual|Bundles are split manually based on specified positions.|Manual bundle splitting serves as a supplementary approach to automatic bundle splitting.<br/><br/>You can use this method in various scenarios, such as: <br/><br/> - If automatic bundle splitting is enabled, but there are still bundles that remain hot for a long time. <br/><br/> - If you want to split bundles and redistribute traffic evenly before having any broker overloaded.

### Workflow

> Below is the workflow for splitting bundles automaticaly or manually.

Author: ok

For example, if the target bundle to split is (0x00000000, 0x80000000), then the bundle split boundary is [0x40000000].

![concept of range_equally_divide](/assets/broker-load-balancing-5.png)

#### topic_count_equally_divide

topic_count_equally_divide splits a bundle into two parts with the same number of topics.

For example, if there are 6 topics in the target bundle [0x00000000, 0x80000000), then you can set the bundle splitting boundary at 0x50000000 to make the left and right sides of the number of topics the same. 

```
hash(topic1) = 0x10000000
hash(topic2) = 0x20000000
hash(topic3) = 0x35000000
hash(topic4) = 0x65000000
hash(topic5) = 0x70000000
hash(topic6) = 0x75000000
```

That is, the target bundle to split is [0x00000000, 0x80000000), and the bundle split boundary is [0x50000000].

![concept of topic_count_equally_divide](/assets/broker-load-balancing-6.png)

For implementation details, see [PR-6241: support evenly distribute topics count when splitting bundles](https://github.com/apache/pulsar/pull/6241).

#### specified_positions_divide

specified_positions_divide splits bundles into several parts by specified positions.

> For example, if you have 2 large topics and there are on the same bundle. Topic1 is on at 0x30000000, Topic2 is on at 0x35000000, and the bundle range is [0x00000000, 0x40000000), then you can set the bundle split boundary as 0x33000000.

Reviewer: ```suggestion
For example, if you have 2 large topics and there are on the same bundle. Topic1 is at 0x30000000, Topic2 is at 0x35000000, and the bundle range is [0x00000000, 0x40000000), then you can set the bundle split boundary as 0x33000000.
```

Author: ok

#### Workflow

Below is the workflow for unloading bundles automatically or manually.

````mdx-code-block
<Tabs groupId="bundle-unloading-workflow"
  defaultValue="Automatic bundle unloading"
  values={[{"label":"Automatic bundle unloading","value":"Automatic bundle unloading"},{"label":"Manual bundle unloading","value":"Manual bundle unloading"}]}>
<TabItem value="Automatic bundle unloading">

#### Step 1: find target brokers

With the broker load information collected from all brokers, the leader broker computes the resource usage of each broker based on the [bundle unloading strategies](tbd).

#### Step 2: unload bundles

If the lead broker finds a broker is overloaded, it will calculate the overloaded bundles, ask the overloaded broker to unload some bundles of topics, remove the target bundles' ownerships, and close the topic sessions and the client connections. 

#### Step 3: assign new owners

The unloaded bundles are assigned to less loaded brokers, and the clients connect to them. 

- For the modular load balancer, bundles will be post-assigned to available brokers when clients send lookup requests.

> - For thhe extensible load balancer, bundles will be pre-assigned to available brokers when unloading.

Reviewer 2: ```suggestion
- For the extensible load balancer, bundles will be pre-assigned to available brokers when unloading.
```

Author: ok
